src/common/servers/serverManager.ts

```diff
@@ -0,0 +1,426 @@
+        }
+        return this.getFirstOpenTabForServer(serverId);
+    }
+
+    getServer = (id: string) => {
+        return this.servers.get(id);
+    }
+
+    getTab = (id: string) => {
+        return this.tabs.get(id);
+    }
+
+    getAllServers = () => {
+        return [...this.servers.values()];
+    }
+
+    hasServers = () => {
+        return Boolean(this.servers.size);
+    }
+
+    getRemoteInfo = (serverId: string) => {
+        return this.remoteInfo.get(serverId);
+    }
+
+    updateRemoteInfos = (remoteInfos: Map<string, RemoteInfo>) => {
+        let hasUpdates = false;
+        remoteInfos.forEach((remoteInfo, serverId) => {
+            this.remoteInfo.set(serverId, remoteInfo);
+            hasUpdates = this.updateServerURL(serverId) || hasUpdates;
+            hasUpdates = this.openExtraTabs(serverId) || hasUpdates;
+        });
+
+        if (hasUpdates) {
+            this.persistServers();
+        }
+    }
+
+    lookupTabByURL = (inputURL: URL | string, ignoreScheme = false) => {
+        log.silly('lookupTabByURL', `${inputURL}`, ignoreScheme);
+
+        const parsedURL = urlUtils.parseURL(inputURL);
+        if (!parsedURL) {
+            return undefined;
+        }
+        const server = this.getAllServers().find((server) => {
+            return equalUrlsIgnoringSubpath(parsedURL, server.url, ignoreScheme) && parsedURL.pathname.match(new RegExp(`^${server.url.pathname}(.+)?(/(.+))?$`));
+        });
+        if (!server) {
+            return undefined;
+        }
+        const tabs = this.getOrderedTabsForServer(server.id);
+
+        let selectedTab = tabs.find((tab) => tab && tab.name === TAB_MESSAGING);
+        tabs.
+            filter((tab) => tab && tab.name !== TAB_MESSAGING).
+            forEach((tab) => {
+                if (parsedURL.pathname.match(new RegExp(`^${tab.url.pathname}(/(.+))?`))) {
+                    selectedTab = tab;
+                }
+            });
+        return selectedTab;
+    }
+
+    updateServerOrder = (serverOrder: string[]) => {
+        log.debug('updateServerOrder', serverOrder);
+
+        this.serverOrder = serverOrder;
+        this.persistServers();
+    }
+
+    updateTabOrder = (serverId: string, tabOrder: string[]) => {
+        log.debug('ServerManager.updateTabOrder', serverId, tabOrder);
+
+        this.tabOrder.set(serverId, tabOrder);
+        this.persistServers();
+    }
+
+    addServer = (server: Team) => {
+        const newServer = new MattermostServer(server, false);
+        this.servers.set(newServer.id, newServer);
+        this.serverOrder.push(newServer.id);
+        const tabOrder: string[] = [];
+        getDefaultTabs().forEach((tab) => {
+            const newTab = this.getTabView(newServer, tab.name, tab.isOpen);
+            this.tabs.set(newTab.id, newTab);
+            tabOrder.push(newTab.id);
+        });
+        this.tabOrder.set(newServer.id, tabOrder);
+
+        this.emit(SERVERS_URL_MODIFIED, [newServer.id]);
```

Comment:
not sure the event being emitted here is actually descriptive, very confusing for someone who hasn't been around the desktop app codebase

Reply:
This event is specific to when a server URL is changed, and basically is there to tell the app to contact the server and update the config information we get from it.

We don't encapsulate this particular logic in the server manager because it's coupled to Electron, and we don't use Electron libraries in the `common` module when we can avoid it.

I'll make a comment to clarify this.